### PR TITLE
allow wasm plugin secret to be read from any namespace

### PR DESCRIPTION
**Pull request by ramaraochavali**
Currently we only allow wasm plugin secret to be read from the same namespace. However since the image secret is same, we can provision it in rootNamespace as well and allow Istio agent to read it from that namespace.

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am not sure I follow why rootNamespace is safe to read from? That is where the most sensitive Secrets live typically

**Comment by ramaraochavali** [comment]
100% agree with the concern. The idea is to have one pull secret instead of the every namespace having to have the same pull secret. Should we fix the name of the secret to be read when rootNamespace is specified?

**Comment by howardjohn** [comment]
It is a good idea but I think it would be good to have:
1. Explicitly allow other namespaces, not implicit rootNamespace logic. IE `istio-system/foo` is ok
2. Explicitly allow permission to the secret, like ReferenceGrant is used for gateway-api to do the same

1 is mostly preference/flexibility, but I think (2) is critical for security

**Comment by ramaraochavali** [comment]
>> Explicitly allow other namespaces, not implicit rootNamespace logic. IE istio-system/foo is ok

We can do that but since rootNamespace is special -  we may be able to allow it to be read from other namespaces.

>>  Explicitly allow permission to the secret, like ReferenceGrant is used for gateway-api to do the same

But  can ReferenceGrant be used here since it is gateway specific? What are other alternatives?

**Comment by howardjohn** [comment]
> > > Explicitly allow other namespaces, not implicit rootNamespace logic. IE istio-system/foo is ok
> 
> We can do that but since rootNamespace is special - we may be able to allow it to be read from other namespaces.
> 
> > > Explicitly allow permission to the secret, like ReferenceGrant is used for gateway-api to do the same
> 
> But can ReferenceGrant be used here since it is gateway specific? What are other alternatives?

I am not sure if there are other reasonable options. There are other projects using ReferenceGrant for non-gateway things now. We definitely need explicit action from the namespace/secret owner to allow folks to access

**Comment by ramaraochavali** [comment]
Changed to use ReferenceGrant. PTAL

**Review by hzxuzhonghu — COMMENTED** [review]
I am not sure we should do this. This is much like pod imagePullSecret

**Comment by ramaraochavali** [comment]
> I am not sure we should do this. This is much like pod imagePullSecret

I think it is not same as pod imagePullSecret. The same WASM filter (for example a standard security filter/JWT filter) can be applied to all workloads in cluster. So this is little different in that sense.

@howardjohn WDYT? Added ReferenceGrant support - PTAL.

**Comment by howardjohn** [comment]
@zirain @ingwonsong ptal

File: pilot/pkg/model/extensions.go
Comment (ingwonsong) [inline pilot/pkg/model/extensions.go:131]: What about passing PushContext directly here, instead of passing a function.

**Comment by ingwonsong** [comment]
IMO, this makes sense a lot. Generally, LGTM.

File: pilot/pkg/model/extensions.go
Comment (ramaraochavali) [reply pilot/pkg/model/extensions.go:131]: That function was added to test this easily without building lot of push context stuff

Comment (ingwonsong) [reply pilot/pkg/model/extensions.go:131]: Yes. the test would be easy, but, IMHO, it seems to be hard to read the codes.

Comment (ingwonsong) [reply pilot/pkg/model/extensions.go:131]: Alternatively, I think we may need to enrich the documentation of `ResourceReferenceAllowed` type.

Comment (ramaraochavali) [reply pilot/pkg/model/extensions.go:131]: I will update the docs to make it more clear

**Comment by hzxuzhonghu** [comment]
Look into the current details again, it automatically make use of secret in the samenamespace as `wasmplugin`
So if the wasm plugin in root namespace and can be applied to all proxies, it has been supported now. So doubt what else need to do?

**Comment by ramaraochavali** [comment]
>  So if the wasm plugin in root namespace and can be applied to all proxies, it has been supported now. So doubt what else need to do?

This is the case of multiple wasm plugins in different namespaces use the same secret. For example if you have wasm filters deploy their images to ECR, all of them will have same secret - So deploying this to all namespaces and managing their life cycle (need to refresh them every 8 hours because these are short lived) is a very complicated process. We are facing this problem now. This allows you to manage this secret in single place (so that it can be refreshed easily) with appropriate reference grants.

**Comment by ramaraochavali** [comment]
> From istio side, to simplify usage. I think we can allow referencing secret from global (root namespace). 

This was my original proposal.  But @howardjohn felt it could be OK to do it for any namespace. I am fine either way

File: pilot/pkg/model/push_context.go
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:2228]: Using SecretAllowed here doesn't seem appropriate. The code is checking for a reference `From KubernetesGateway -> To Secret`. But this is WASM -- its not referenced from a KubernetesGateway but a WASMPlugin

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:2228]: I think - double checking with gateway folks

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/model/push_context.go
Comment (ramaraochavali) [reply pilot/pkg/model/push_context.go:2228]: You are right. I did not look at the implementation, It is hard coding kubernetes gateway.,

Comment (ramaraochavali) [reply pilot/pkg/model/push_context.go:2228]: Also just realized we do not process any other reference grants other than gateway resources [here](https://github.com/istio/istio/blob/5a1314991c1daef5249888015fff0e0c46eda03e/pilot/pkg/config/kube/gateway/conversion.go#L184). So if we go down this path this needs to be enhanced as well.

**Comment by ramaraochavali** [comment]
@howardjohn @hzxuzhonghu WDYT? Should I continue with this(fixing the current stuff) or you have concerns with this approach?

**Comment by hzxuzhonghu** [comment]
Yes, i think there is security concern about referencing secret from other namespace.

**Comment by howardjohn** [comment]
I personally think its ok with ReferenceGrant, which requires users to explicitly allow access to the secret

**Comment by hzxuzhonghu** [comment]
After searching found https://github.com/kubernetes/ingress-nginx/issues/2170
And also https://github.com/istio/istio/issues/19970#issuecomment-634849254 @howardjohn

**Comment by howardjohn** [comment]
This was precisely what reference grant was created for and how we are able to securely implement cross-namespace access in k8s gateway. Note my comment is from 3 years ago before gateway-api solved this with reference grant

**Comment by hzxuzhonghu** [comment]
I could see most of the workaround is to replicate namespace resources.  Unless `reference grant` is from k/k upstream, otherwise I do not think it is recommended way.

**Comment by howardjohn** [comment]
reference grant is currently part of gateway-api but it's moving to core actually. I'll try to find the KEP link

**Comment by ramaraochavali** [comment]
@hzxuzhonghu gentle ping. Are you OK with using reference grants? If yes, I will update the PR

**Comment by hzxuzhonghu** [comment]
I feel not do that until it goes into k8s core api

**Comment by ramaraochavali** [comment]
> I feel not do that until it goes into k8s core api

@howardjohn can you please share the KEP? Without supporting wasm image pull secret in root namespace - I think WasmPlugin is pretty unusable. We have standard set of security filters that we want to apply it to every sidecar and imagine if we have to rotate the pull secret (as most of the clouds like AWS have short expiry time ~ 8hrs) in every namespace. I think it is a huge operational overhead.

**Comment by howardjohn** [comment]
https://github.com/kubernetes/enhancements/issues/3766

**Comment by ingwonsong** [comment]
If this issue is urgent for some use cases, as a short-term approach, we may need to consider again about sharing implicitly the secret in `istio-system` namespace. Since the namespace is just for Istio, I don't think it has an issue.

WDYT?

@howardjohn @hzxuzhonghu

**Comment by howardjohn** [comment]
istio-system is the one namespace that we *know* has highly privileged
Secrets (ca key), so I don't think that is a safe idea.

**Comment by ingwonsong** [comment]
Ah

> istio-system is the one namespace that we *know* has highly privileged Secrets (ca key), so I don't think that is a safe idea.

Ah.. right. That's good point.

**Comment by ramaraochavali** [comment]
This PR started with treating "istio-system" as special namespace and allowing secret reads from it. But I agree with @howardjohn that it is risky to do that. I can not think of any alternatives except for using the reference grants.

**Comment by hzxuzhonghu** [comment]
As the kep has been merged, it seems not too long before the api can go into k/k repo.

**Comment by ramaraochavali** [comment]
Should we wait for it to be merged or can we go ahead with this?

**Comment by ramaraochavali** [comment]
@hzxuzhonghu WDYT ^^? Assuming it gets approved, should we move forward with reference grants?

**Comment by hzxuzhonghu** [comment]
Prefer waiting until it get landed, because it will be in different api group

**Pull request closed without merging**
